File: python/01_pronunciation_assessment.py
"""
Pronunciation Assessment — Vocametrix API example

Scores how accurately a speaker pronounces a reference phrase.
Returns phoneme-level accuracy scores for 30+ locales.

Upload pattern: get-blob-url → PUT to Azure → POST /api/pronunciation-assessment
Auth: X-API-Key header

Usage:
    python 01_pronunciation_assessment.py path/to/audio.wav
    python 01_pronunciation_assessment.py path/to/audio.wav --text "Hello world" --locale en-US
"""

import sys
import os
import json
import argparse
import requests
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}


def upload_via_blob_url(audio_path: str) -> str:
    """Upload audio through the Azure blob-url pattern, return blobURL."""
    r = requests.post(f'{BASE_URL}/api/get-blob-url', headers=HEADERS)
    r.raise_for_status()
    data = r.json()
    upload_url = data['uploadURL']
    blob_url = data['blobURL']

    with open(audio_path, 'rb') as f:
        put = requests.put(
            upload_url,
            data=f,
            headers={'x-ms-blob-type': 'BlockBlob', 'Content-Type': 'audio/wav'},
        )
        put.raise_for_status()

    return blob_url


def assess_pronunciation(audio_path: str, text: str, locale: str) -> dict:
    blob_url = upload_via_blob_url(audio_path)

    r = requests.post(
        f'{BASE_URL}/api/pronunciation-assessment',
        headers=HEADERS,
        json={'blobURL': blob_url, 'referenceText': text, 'locale': locale},
    )
    r.raise_for_status()
    return r.json()


def print_results(results: dict):
    print('\n=== Pronunciation Assessment Results ===')
    print(f"Overall accuracy:    {results.get('accuracyScore', 'N/A'):.1f} / 100")
    print(f"Fluency score:       {results.get('fluencyScore', 'N/A'):.1f} / 100")
    print(f"Completeness score:  {results.get('completenessScore', 'N/A'):.1f} / 100")
    print(f"Pronunciation score: {results.get('pronunciationScore', 'N/A'):.1f} / 100")

    words = results.get('words', [])
    if words:
        print('\nWord-level breakdown:')
        for w in words:
            print(f"  {w.get('word', '?'):20s}  accuracy={w.get('accuracyScore', '?'):.0f}")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Pronunciation assessment')
    parser.add_argument('audio', help='Path to WAV file')
    parser.add_argument('--text', default='Hello, my name is Alex.', help='Reference text')
    parser.add_argument('--locale', default='en-US', help='Language locale (e.g. en-US, fr-FR)')
    args = parser.parse_args()

    results = assess_pronunciation(args.audio, args.text, args.locale)
    print_results(results)

File: javascript/01-pronunciation-assessment.js
/**
 * Pronunciation Assessment — Vocametrix API example (Node.js)
 *
 * Scores how accurately a speaker pronounces a reference phrase.
 * Returns phoneme-level accuracy scores for 30+ locales.
 *
 * Upload pattern: get-blob-url → PUT to Azure → POST /api/pronunciation-assessment
 * Auth: X-API-Key header
 *
 * Usage:
 *   node 01-pronunciation-assessment.js path/to/audio.wav
 *   node 01-pronunciation-assessment.js audio.wav --text "Hello world" --locale en-US
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';
const HEADERS = { 'X-API-Key': API_KEY, 'Content-Type': 'application/json' };

async function uploadViaBlobUrl(audioPath) {
  const res = await fetch(`${BASE_URL}/api/get-blob-url`, { headers: HEADERS });
  if (!res.ok) throw new Error(`get-blob-url failed: ${await res.text()}`);
  const { uploadURL, blobURL } = await res.json();

  const audioData = fs.readFileSync(audioPath);
  const put = await fetch(uploadURL, {
    method: 'PUT',
    body: audioData,
    headers: { 'x-ms-blob-type': 'BlockBlob', 'Content-Type': 'audio/wav' },
  });
  if (!put.ok) throw new Error(`Azure upload failed: ${await put.text()}`);
  return blobURL;
}

async function assessPronunciation(audioPath, text, locale) {
  const blobURL = await uploadViaBlobUrl(audioPath);
  const res = await fetch(`${BASE_URL}/api/pronunciation-assessment`, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify({ blobURL, referenceText: text, locale }),
  });
  if (!res.ok) throw new Error(`pronunciation-assessment failed: ${await res.text()}`);
  return res.json();
}

function printResults(results) {
  console.log('\n=== Pronunciation Assessment Results ===');
  console.log(`Overall accuracy:    ${results.accuracyScore?.toFixed(1) ?? 'N/A'} / 100`);
  console.log(`Fluency score:       ${results.fluencyScore?.toFixed(1) ?? 'N/A'} / 100`);
  console.log(`Completeness score:  ${results.completenessScore?.toFixed(1) ?? 'N/A'} / 100`);
  console.log(`Pronunciation score: ${results.pronunciationScore?.toFixed(1) ?? 'N/A'} / 100`);

  const words = results.words ?? [];
  if (words.length > 0) {
    console.log('\nWord-level breakdown:');
    for (const w of words) {
      console.log(`  ${w.word?.padEnd(20) ?? '?'}  accuracy=${w.accuracyScore?.toFixed(0) ?? '?'}`);
    }
  }
}

// --- CLI ---
const args = process.argv.slice(2);
if (args.length === 0) {
  console.error('Usage: node 01-pronunciation-assessment.js <audio.wav> [--text "..."] [--locale en-US]');
  process.exit(1);
}
const audioPath = args[0];
const textIdx = args.indexOf('--text');
const localeIdx = args.indexOf('--locale');
const text = textIdx !== -1 ? args[textIdx + 1] : 'Hello, my name is Alex.';
const locale = localeIdx !== -1 ? args[localeIdx + 1] : 'en-US';

assessPronunciation(audioPath, text, locale)
  .then(printResults)
  .catch(err => { console.error(err.message); process.exit(1); });
